### PR TITLE
 Sixth iteration of the Court pallet 

### PR DESCRIPTION
**Pull request by c410-f3r**
Depends on #233 
cc #200 

Basically brings back another bunch of storage and types from `simple-disputes` to `PM` since they are used on both `court` and `simple-disputes` pallets

### CODE REVIEW

**Review by sea212 — APPROVED** [review]
Thanks, looks good!

File: zrml/prediction-markets/src/lib.rs [hunk 1]
```diff
@@ -970,6 +945,17 @@ mod pallet {
...
+    /// A market only ends up here if it was disputed.
+    #[pallet::storage]
+    pub type MarketIdsPerDisputeBlock<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::BlockNumber, Vec<MarketIdOf<T>>, ValueQuery>;
+
+    /// A mapping of market identifiers to the block that they were reported on.
+    #[pallet::storage]
+    pub type MarketIdsPerReportBlock<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::BlockNumber, Vec<MarketIdOf<T>>, ValueQuery>;
+
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply zrml/prediction-markets/src/lib.rs:958]: Changed

File: zrml/prediction-markets/src/lib.rs [hunk 2]
```diff
@@ -1161,6 +1147,133 @@ mod pallet {
...
 
+        // If a market is categorical, destroys all non-winning assets.
+        fn manage_resolved_categorical_market(
+            market: &Market<T::AccountId, T::BlockNumber>,
+            market_id: &MarketIdOf<T>,
+            outcome_report: &OutcomeReport,
+        ) -> Result<[usize; 3], DispatchError> {
+            let mut total_accounts: usize = 0;
+            let mut total_asset_accounts: usize = 0;
+            let mut total_categories: usize = 0;
```
[the remark this answers is not in the document]
Comment (c410-f3r) [reply zrml/prediction-markets/src/lib.rs:1158]: Although Substrate will probably never run on an 8 bit micro-controller, the possible effects of indiscriminate casts might appear in the future as silent bugs or as a blocker with an hypothetical `wasm64` target.
I personally prefer to work exclusively with `usize` unless a strong reason tells otherwise.

[the remark this answers is not in the document]
Comment (pepyakin) [reply zrml/prediction-markets/src/lib.rs:1158]: Hey, found your discussion by the backlink in the OG issue. Just wanted to chime in and give my perspective on wasm32 vs. wasm64.

There are several things in play. 

We generally avoid using `usize`. This is not exclusively related to Substrate Runtime. For example, parity-scale-codec does not provide an impl for `usize` just because it is so easy to get similar problems e.g. on arm (32bit) vs aarch64 vs x86 vs x86_64.

But yes, Wasm Runtime assumes a 32 bit word. wasm64 is on our radar as well and so far it seems we are not going to leverage it. This is because wasm64 only affects memory addressing. It is also doesn't make you use one or another and it is theoretically possible to mix 32 and 64 memories within a single instance. Basically moving from wasm32 to wasm64 will allow us to use a wider pointer. This allows to address more memory but it also leads to more memory consumption, which in turn hurts performance through less optimal CPU cache usage. At the same time, I do not think we want or even can afford to use 4 GiB per wasm instance, and I do not see us relying on virtual memory for tricks, and I do not think tagged pointers or all other stuff gives us much.

Bottom line, I think we can safely assume that Wasm Runtime will stay on wasm32. Sorry if this is too off topic for this place feel free to move this dicussion elsewhere.

**Review by lsaether — APPROVED** [review]

File: zrml/prediction-markets/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (c410-f3r) [reply zrml/prediction-markets/src/lib.rs:1158]: Hello @apopiak 

It is really to see someone from Parity caring about this subject, the more the Polkadot ecosystem unites to help each other, the more successful it will be. Maybe the same attention can no be applied to my Substrate PRs or my unanswered Element questions but that is another history :)

Thanks for asserting that `usize` will probably always be `u32`, this brings more confidence to our pallet development and please, feel free to chime in any time! Your comments and expertise are of great value!
